services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java

```diff
@@ -165,6 +167,6 @@ private void handleLimitExceeded(AuthenticationFlowContext context, List<UserSes
     private void logoutOldestSession(List<UserSessionModel> userSessions) {
         logger.info("Logging out oldest session");
         Optional<UserSessionModel> oldest = userSessions.stream().sorted(Comparator.comparingInt(UserSessionModel::getLastSessionRefresh)).findFirst();
-        oldest.ifPresent(userSession -> AuthenticationManager.backchannelLogout(session, userSession, true));
+        oldest.ifPresent(userSession -> AuthenticationManager.backchannelLogout(session, lockObjectsForModification(session, () -> session.sessions().getUserSession(userSession.getRealm(), userSession.getId())), true));
```

Comment:
To be able to lock the user session (which is going to be removed) the provider is called even the `UserSessionModel` is already available. Other alternative would be to lock all user sessions from the realm (line 56), but it doesn't seem like a correct solution.

Reply:
Here, I have the same comment as below: Locking of rows in the database works only when the entity is read from the database in the session for the first time. When it is locked later, the session might already have a stale entry in its persistence context.

Here, the list of session is collected in line 56 with `session.sessions().getUserSessionsStream(context.getRealm(), context.getUser())`. IMHO this is the call that should be wrapped with `lockObjectsForModification()`, afterwards no locking will be necessary. 

I wonder if locking was necessary here to make a test pass? Without locking, this could lead to exceptions when a user tries to log in simultaneously, and their oldest session is beging logged out concurrently. 

When this would lock the whole list of the user's sessions in line 56, this would ensure that no concurrent counting of those sessions happens, and that the simultaneous log in would wait until the first evaluation has completed. Using locking in this way will lead to consistent outcomes, and no concurrency problems.

Reply:
As discussed in a call - locking was moved to line 56.